torch/fx/graph_module.py

```diff
@@ -642,12 +642,50 @@ def code(self) -> str:
         Return the Python code generated from the ``Graph`` underlying this
         ``GraphModule``.
         """
+
+        self.real_recompile()
         if not hasattr(self, '_code'):
             raise RuntimeError('Code has not been generated! Please report a bug to PyTorch')
         return self._code
 
+    @property
+    def in_spec(self):
+        self.real_recompile()
+        return getattr(self, "_in_spec", None)
+
+    @property
+    def out_spec(self):
+        self.real_recompile()
+        return getattr(self, "_out_spec", None)
+
     @compatibility(is_backward_compatible=True)
-    def recompile(self) -> PythonCode:
+    @classmethod
+    def recompile(cls):
+        cls.forward = cls._lazy_forward
+
+    @classmethod
+    def _needs_recompile(cls):
+        return cls.forward is cls._lazy_forward
+
+    def _lazy_forward(self, *args, **kwargs):
+        self._real_recompile()
+        assert not self._needs_recompile()
+        return self.forward(*args, **kwargs)
+
+    forward = _lazy_forward
+
+    def real_recompile(self):
+        """
+        A torch script safe wrapper around _real_recompile.
```

Comment:
It's incredibly strange to me that recompile is being called from TorchScriptable code!

Reply:
Here is how it happens from looking at some failure stack trace. When torchscript gets a GraphModule, it will try to inspect the GraphModule's attributes. When it looks at the 'code' property, it will try to 'script' the method implementation. Due to the lazy recompilation, we may need to recompile the graph module when 'code' property is being accessed. That's how what you described happens.

I think a not too bad fix is to always force a ***real*** recompilation when TorchScript backend gets a graph module.

Reply:
Ideally yes. But this hits some limitation of torchscript - some of the lazy recompile related methods can not be handled by torchscript. Since we are deprecating torchscript, probably we should workaround rather than spending efforts to improve torchscript?

Reply:
@ezyang  so we encounter similar issues when passing a lazy recompiled GraphModule to dynamo. Similar to the torchscript case, there is an option to enhance the graph capturing layer (dynamo or torchscript) to handle the GraphModule lazy recompilation. But that may take quite some effort and I'm not sure if it's necessary. it would be handy to just force the recompile at the beginning of these graph capturing layers. Wdyt?